[Ready] Adds 5 new negative/minor negative mutations, mutation conflict system

PKPenguin321 opened the pull request:
:cl:
add: New system in genetics: Conflicting mutations. Some mutations will not allow others to activate. Example - If you have mutation A, and it conflicts with B, then B will never activate until you get rid of mutation A (and visa versa). The only mutations to currently use this are Dwarfism and Gigantism. Speaking of which...
add: New minor negative mutation: Gigantism. Opposite of dwarfism, makes you larger.
add: New disability: Spastic. You will randomly take a step in a random direction, will randomly change intent, and will randomly try to click on somebody near you (which can hug them, punch them, use whatever is in your hand on them, etc).
add: New disability: Acid flesh. Pockets of acid bubble beneath your skin. They will occasionally pop, coating you with acid. This will melt clothes and deal damage to you.
add: New disability: Spatial Instability. You will randomly teleport, similar to a wizard's blink teleport. This is a very nauseating process, and will inevitably make you very sick if you don't rid yourself of it.
add: New disability: Two Left Feet. Your right foot becomes a second left foot. You take longer to get up from stuns.
/:cl:

Will probably add combination-only powers that use these as components in future PRs.

PKPenguin321 committed:
adds 5 new (minor) negative mutations
also adds a mutation conflict system

Razharas commented:
A general solution to this is recipe system
With recipe resolving you can have smooth system with tricky cases like mutations X and Y are conflicting but when you get new one the old one is displaced
Or Mutation X always overrides mutation Y but not the other way
Or custom things like "you can only have any 4 of these 6 mutation types" or "you can only get this mutation if you have these 2 others"

PKPenguin321 commented:
Until there are mutations that would actually benefit from a system like that, I see no point. Plus I think that powers should prevent other powers when they conflict rather than be overwritten, because if they were to overwrite you could just swap powers on the fly with injectors.

PKPenguin321 committed:
removes debug comments and to_chat(world)s

PKPenguin321 reviewed (commented):
notes to self

coiax commented on code/datums/mutations/body.dm:
```diff
@@ -230,6 +231,108 @@
+	var/warpchance = 0
+
+/datum/mutation/human/badblink/on_life()
+	if(prob(warpchance))
```
Disable this if they are inside something else, like a cloning pod or cryo tube. There were issues with the Void mutation where someone developed it inside a cloning pod, ejected themselves early, but still had all the cloning pod traits attached to them.

PKPenguin321 replied:
You sure do_teleport doesn't handle this already? Guess I'll test for it

Shadowflame909 commented:
Negative Mutations lower instability right? Allowing you to take more powers with the more negative ones you have?

PKPenguin321 commented on code/datums/mutations/body.dm:
```diff
@@ -230,6 +231,108 @@
+	text_lose_indication = "<span class'notice'>Your flinching subsides.</span>"
+	difficulty = 16
+
+/datum/mutation/human/spastic/on_life()
```
I'll have to look into it, but frankly that sounds a lot shakier than what I've already written

ExcessiveUseOfCobblestone replied:
If it already exists modularize it to fit here and perfect the current version vs. having 2 different versions that do the same thing.

PKPenguin321 commented:
Good question. I believe it shouldn't affect that but I could be wrong, haven't looked at the code for that

Time-Green commented:
Looks like it doesn’t discover judging by code. I guess not having your subject have alot of mutations might be a good solution, because bypassing this will be a pain codewise

SouDescolado commented:
I assume Spatic ignores Pacifism when punching people?

SouDescolado commented:
Also it would be funny if every time someone got cured from a mutation that has a conflict, there is a chance (maybe low, maybe 50) it gets the conflicting mutation. Like.. Had Gigantism? You are now Dwarf.

PKPenguin321 commented:
>I assume Spastic ignores Pacifism when punching people?

Not sure if there's a clean way to do that due to the way pacifism works, sadly

ExcessiveUseOfCobblestone commented on code/datums/mutations.dm:
```diff
@@ -47,6 +48,13 @@
+		if(LAZYLEN(mewtayshun.conflicts))
+			for(var/cons in mewtayshun.conflicts)
+				var/datum/mutation/human/conflicter = cons
+				if(conflicter == src.type)
```
```suggestion
				if(conflicter == type)
```

ExcessiveUseOfCobblestone commented on code/datums/mutations/body.dm:
```diff
@@ -230,6 +231,108 @@
+
+/datum/mutation/human/acidflesh/on_life()
+	if(prob(25))
+		to_chat(owner, "<span class='danger'>Your acid flesh bubbles...</span>")
```
does this not spam chat?

PKPenguin321 replied:
Intended, presses urgency and it squashes anyways. Not as frequent as you might expect either.

ExcessiveUseOfCobblestone replied:
It doesn't squash when you have a bunch of players chatting in radio or doing literally anything else you memer.

It needs a cooldown or something to ensure it RNG won't clusterheck your chat panel.

PKPenguin321 replied:
Ok, I put a 20 second minimum cooldown

ExcessiveUseOfCobblestone commented on code/datums/mutations/body.dm:
```diff
@@ -230,6 +231,108 @@
+	if(prob(15))
+		var/arm_and_a_leg = rand(0,1)
 
+		if(arm_and_a_leg)
```
you want prob(50) here.

ExcessiveUseOfCobblestone commented on code/datums/mutations/body.dm:
```diff
@@ -230,6 +231,108 @@
 
+/datum/mutation/human/badblink
+	name = "Spatial Instability"
+	desc = "Subject has a very weak link to spatial reality, and may be displaced. Often causes extreme nausea."
```
Just FYI the mutation desc usually has the mutation itself as the subject rather than the person which manifests such a mutation. It's really triggering the perfectionist within.

ExcessiveUseOfCobblestone commented on code/datums/mutations/body.dm:
```diff
@@ -230,6 +231,108 @@
 
+/datum/mutation/human/extrastun
+	name = "Two Left Feet"
+	desc = "Subject literally has two left feet, making standing up after getting knocked down very difficult."
```
```suggestion
	desc = "A mutation that replaces the right foot with another left foot. Scientists are still debating whether such a mutation makes for good stand-up or not."
```

ExcessiveUseOfCobblestone commented on code/datums/mutations/body.dm:
```diff
@@ -230,6 +231,108 @@
+	text_gain_indication = "<span class='warning'>Your right foot feels... left.</span>"
+	text_lose_indication = "<span class'notice'>Your right foot feels alright.</span>"
+	difficulty = 16
+	var/stun_cooldown = 0
```
reeeeeee

PKPenguin321 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into genetics-negatives-bundle

# Conflicts:
#	code/__DEFINES/DNA.dm
#	code/game/objects/items/dna_injector.dm

PKPenguin321 committed:
addresses reviews

PKPenguin321 committed:
further review addressing, new message
and a message given to the subject when a mutation fails to manifest because of conflicts

PKPenguin321 commented:
Should be good to go. Tested on a local server

PKPenguin321 committed:
cooldown on acid flesh message

adamkad1 commented:
So are disabilities gonna increase stability some time?

Time-Green commented:
maybe
